File: map_lib/test/test_single.cpp
#include <gmock/gmock.h>

#include <map_lib/type/cone.hpp>
#include <map_lib/type/map.hpp>

#include "map_lib/association/single.hpp"
#include "map_lib/association/filter/thresholding.hpp"
#include "map_lib/association/metric/euclidean.hpp"

#include "make_types.hpp"  // NOLINT

using eufs::map::association::Euclidean;
using eufs::map::association::SingleAssociation;
using eufs::map::association::Threshold;
using eufs::map::Map;
using eufs::map::Cone;

///////////////////
//  Check Build  //
///////////////////

TEST(CheckBuild, Build) { SingleAssociation<Threshold<Euclidean>, Euclidean>(1.0); }

/////////////////
// Check Match //
/////////////////

TEST(CheckMatch, SingleObservationTwoValidOptions) {
  std::vector<Cone> observations{MakeCone({1.0, 1.2}, {0.1, 0.0, 0.0, 0.1})};
  Map<Cone> map{std::vector<Cone>{MakeCone({0.5, 1.3}, {0.4, 0.0, 0.0, 0.4}),
                                  MakeCone({1.0, 1.1}, {0.4, 0.0, 0.0, 0.4})}};
  auto association = SingleAssociation<Threshold<Euclidean>, Euclidean>(1.0);
  auto res = association.Match(observations, map);
  EXPECT_THAT(res.matched, ::testing::UnorderedElementsAreArray({::testing::Pair(0lu, 1lu)}));
  EXPECT_THAT(res.unmatched.size(), 0lu);
}

TEST(CheckMatch, SingleObservationOneValidOption) {
  std::vector<Cone> observations{MakeCone({1.0, 1.2}, {0.1, 0.0, 0.0, 0.1})};
  Map<Cone> map{std::vector<Cone>{MakeCone({0.5, 1.3}, {0.4, 0.0, 0.0, 0.4}),
                                  MakeCone({1.0, 1.1}, {0.4, 0.0, 0.0, 0.4})}};
  auto association = SingleAssociation<Threshold<Euclidean>, Euclidean>(0.25);
  auto res = association.Match(observations, map);
  EXPECT_THAT(res.matched, ::testing::UnorderedElementsAreArray({::testing::Pair(0lu, 1lu)}));
  EXPECT_THAT(res.unmatched.size(), 0lu);
}

TEST(CheckMatch, SingleObservationNoValidOption) {
  std::vector<Cone> observations{MakeCone({1.5, 2.0}, {0.1, 0.0, 0.0, 0.1})};
  Map<Cone> map{std::vector<Cone>{MakeCone({0.5, 1.3}, {0.4, 0.0, 0.0, 0.4}),
                                  MakeCone({1.0, 1.1}, {0.4, 0.0, 0.0, 0.4})}};
  auto association = SingleAssociation<Threshold<Euclidean>, Euclidean>(0.25);
  auto res = association.Match(observations, map);
  EXPECT_THAT(res.matched.size(), 0lu);
  EXPECT_THAT(res.unmatched, ::testing::UnorderedElementsAreArray({0lu}));
}

TEST(CheckMatch, UnmatchedObservation) {
  std::vector<Cone> observations{
      MakeCone({1.0, 1.2}, {0.1, 0.0, 0.0, 0.1}),
      MakeCone({5.9, 1.2}, {0.1, 0.0, 0.0, 0.1}),
  };
  Map<Cone> map{std::vector<Cone>{
      MakeCone({0.5, 1.3}, {0.4, 0.0, 0.0, 0.4}),
      MakeCone({1.0, 1.1}, {0.4, 0.0, 0.0, 0.4}),
      MakeCone({5.0, 1.1}, {0.4, 0.0, 0.0, 0.4}),
  }};
  auto association = SingleAssociation<Threshold<Euclidean>, Euclidean>(0.25);
  auto res = association.Match(observations, map);
  EXPECT_THAT(res.matched, ::testing::UnorderedElementsAreArray({
                               ::testing::Pair(0lu, 1lu),
                           }));
  EXPECT_THAT(res.unmatched, ::testing::UnorderedElementsAreArray({1lu}));
}

TEST(CheckMatch, DuplicateAssociation) {
  std::vector<Cone> observations{
      MakeCone({1.0, 1.2}, {0.1, 0.0, 0.0, 0.1}),
      MakeCone({0.9, 1.2}, {0.1, 0.0, 0.0, 0.1}),
  };
  Map<Cone> map{std::vector<Cone>{
      MakeCone({0.5, 1.3}, {0.4, 0.0, 0.0, 0.4}),
      MakeCone({1.0, 1.1}, {0.4, 0.0, 0.0, 0.4}),
      MakeCone({5.0, 1.1}, {0.4, 0.0, 0.0, 0.4}),
  }};
  auto association = SingleAssociation<Threshold<Euclidean>, Euclidean>(0.25);
  auto res = association.Match(observations, map);
  EXPECT_THAT(res.matched, ::testing::UnorderedElementsAreArray({
                               ::testing::Pair(0lu, 1lu),
                               ::testing::Pair(1lu, 1lu),
                           }));
  EXPECT_THAT(res.unmatched.size(), 0lu);
}

File: map_lib/test/test_single.py
from map_lib.association import SingleAssociation
from map_lib.type import Cone


class TestSingle:
    def test_constructor(self):
        SingleAssociation(1.0)

    def test_single_observation_two_valid_options(self):
        observations = [Cone([1.0, 1.2], [0.1, 0.0, 0.0, 0.1])]
        cone_map = [Cone([0.5, 1.3], [0.4, 0.0, 0.0, 0.4]),
                    Cone([1.0, 1.1], [0.4, 0.0, 0.0, 0.4])]
        association = SingleAssociation(1.0)
        res = association.match(observations, cone_map)
        assert res.matched == {0: 1}
        assert len(res.unmatched) == 0

    def test_single_observation_one_valid_option(self):
        observations = [Cone([1.0, 1.2], [0.1, 0.0, 0.0, 0.1])]
        cone_map = [Cone([0.5, 1.3], [0.4, 0.0, 0.0, 0.4]),
                    Cone([1.0, 1.1], [0.4, 0.0, 0.0, 0.4])]
        association = SingleAssociation(0.25)
        res = association.match(observations, cone_map)
        assert res.matched == {0: 1}
        assert len(res.unmatched) == 0

    def test_single_observation_no_valid_option(self):
        observations = [Cone([1.5, 2.0], [0.1, 0.0, 0.0, 0.1])]
        cone_map = [Cone([0.5, 1.3], [0.4, 0.0, 0.0, 0.4]),
                    Cone([1.0, 1.1], [0.4, 0.0, 0.0, 0.4])]
        association = SingleAssociation(0.25)
        res = association.match(observations, cone_map)
        assert len(res.matched) == 0
        assert res.unmatched == {0}

    def test_unmatched_observations(self):
        observations = [
            Cone([1.0, 1.2], [0.1, 0.0, 0.0, 0.1]),
            Cone([5.9, 1.2], [0.1, 0.0, 0.0, 0.1])
        ]
        cone_map = [
            Cone([0.5, 1.3], [0.4, 0.0, 0.0, 0.4]),
            Cone([1.0, 1.1], [0.4, 0.0, 0.0, 0.4]),
            Cone([5.0, 1.1], [0.4, 0.0, 0.0, 0.4])
        ]
        association = SingleAssociation(0.25)
        res = association.match(observations, cone_map)
        assert res.matched == {0: 1}
        assert res.unmatched == {1}

    def test_duplicate_association(self):
        observations = [
            Cone([1.0, 1.2], [0.1, 0.0, 0.0, 0.1]),
            Cone([0.9, 1.2], [0.1, 0.0, 0.0, 0.1])
        ]
        cone_map = [
            Cone([0.5, 1.3], [0.4, 0.0, 0.0, 0.4]),
            Cone([1.0, 1.1], [0.4, 0.0, 0.0, 0.4]),
            Cone([5.0, 1.1], [0.4, 0.0, 0.0, 0.4])
        ]
        association = SingleAssociation(0.25)
        res = association.match(observations, cone_map)
        assert res.matched == {0: 1, 1: 1}
        assert len(res.unmatched) == 0
